aesd-char-driver/main.c

```diff
@@ -18,58 +18,202 @@
 #include <linux/cdev.h>
 #include <linux/fs.h> // file_operations
 #include "aesdchar.h"
+#include "aesd-circular-buffer.h"
+#include <linux/slab.h>
+
 int aesd_major =   0; // use dynamic major
 int aesd_minor =   0;
 
-MODULE_AUTHOR("Your Name Here"); /** TODO: fill in your name **/
+MODULE_AUTHOR("Deekshith Reddy Patil"); /** TODO: fill in your name **/
 MODULE_LICENSE("Dual BSD/GPL");
 
 struct aesd_dev aesd_device;
 
 int aesd_open(struct inode *inode, struct file *filp)
 {
 	PDEBUG("open");
+
 	/**
 	 * TODO: handle open
 	 */
+
 	return 0;
 }
 
 int aesd_release(struct inode *inode, struct file *filp)
 {
+
 	PDEBUG("release");
 	/**
 	 * TODO: handle release
 	 */
+
 	return 0;
 }
 
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-	ssize_t retval = 0;
+	struct aesd_buffer_entry *entry;
+	size_t entry_offset_byte_rtn;
+
+	int bytes_copied = 0;
+	int this_read = 0;
+	loff_t initial_fpos;
+
 	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
 	/**
 	 * TODO: handle read
 	 */
-	return retval;
+
+	mutex_lock(&aesd_device.aesd_mutex);
+
+	initial_fpos = *f_pos;
+
+	while(bytes_copied  < count)
+	{
+		entry = aesd_circular_buffer_find_entry_offset_for_fpos(&aesd_device.aesd_circular_buffer,*f_pos,&entry_offset_byte_rtn);
+
+		if(entry == NULL)
+		{
+			*f_pos += bytes_copied;
+
+			mutex_unlock(&aesd_device.aesd_mutex);
+
+			PDEBUG("Returning in between, bytes_copied = %d\n",bytes_copied);
+			return bytes_copied;
+		}
+
+		this_read = entry->size - entry_offset_byte_rtn;
+
+		if(copy_to_user(&buf[bytes_copied],&entry->buffptr[entry_offset_byte_rtn],this_read))
+    	{
+			mutex_unlock(&aesd_device.aesd_mutex);
```

Comment:
@DeekshithPatil add common graceful return at the end rather than copying the code.